src/processor/graph-update-completion-monitor.service.ts

```diff
@@ -0,0 +1,124 @@
+      const moduleError = registry.findMetaError(moduleThatErrored);
+      newStatus.error = moduleError.method;
+      newStatus.status = 'failed';
+    } else if (hasSuccess) {
+      newStatus.status = 'success';
+    } else {
+      this.logger.error(`Tx hash ${newStatus.txHash} found in block, but neither success nor failure`);
+      // No change in status
+    }
+
+    return newStatus;
+  }
+
+  async processCurrentBlock(currentBlockHash: BlockHash, currentBlockNumber: number): Promise<void> {
+    // Get set of tx hashes to monitor from cache
+    let pendingTxns = await this.cacheService.getAllPendingTxns();
+
+    const block = await this.blockchainService.getBlock(currentBlockHash);
+    const extrinsicIndices: [HexString, number][] = [];
+    block.block.extrinsics.forEach((extrinsic, index) => {
+      if (Object.keys(pendingTxns).some((txHash) => txHash === extrinsic.hash.toHex())) {
+        extrinsicIndices.push([extrinsic.hash.toHex(), index]);
+      }
+    });
+
+    if (extrinsicIndices.length > 0) {
+      const at = await this.blockchainService.apiPromise.at(currentBlockHash);
+      const epoch = (await at.query.capacity.currentEpoch()).toNumber();
+      const events = (await at.query.system.events()).filter(({ phase }) => phase.isApplyExtrinsic && extrinsicIndices.some((index) => phase.asApplyExtrinsic.eq(index)));
+
+      const totalCapacityWithdrawn: bigint = events
+        .filter(({ event }) => at.events.capacity.CapacityWithdrawn.is(event))
+        .reduce((sum, { event }) => (event as unknown as any).data.amount.toBigInt() + sum, 0n);
+
+      const updateTxns: ITxStatus[] = [];
+
+      // eslint-disable-next-line no-restricted-syntax
+      for (const [txHash, txIndex] of extrinsicIndices) {
+        const extrinsicEvents = events.filter(({ phase }) => phase.isApplyExtrinsic && phase.asApplyExtrinsic.eq(txIndex));
+        const hasSuccess = extrinsicEvents.some(({ event }) => at.events.utility.BatchCompleted.is(event));
+        const failureEvent = extrinsicEvents.find(({ event }) => at.events.system.ExtrinsicFailed.is(event))?.event;
+
+        updateTxns.push(this.getTxStatus(pendingTxns[txHash], hasSuccess, failureEvent));
+      }
+      await this.cacheService.upsertWatchedTxns(updateTxns);
+
+      await this.setEpochCapacity(epoch, totalCapacityWithdrawn);
+
+      // Now check all pending transactions for expiration as of this block
+      pendingTxns = await this.cacheService.getAllPendingTxns();
+      // eslint-disable-next-line no-restricted-syntax
+      for (const txStatus of Object.values(pendingTxns)) {
+        if (txStatus.death <= currentBlockNumber) {
+          txStatus.status = 'expired';
+          this.logger.verbose(`Tx ${txStatus.txHash} expired (birth: ${txStatus.birth}, death: ${txStatus.death}, currentBlock: ${currentBlockNumber})`);
+          // eslint-disable-next-line no-await-in-loop
+          await this.cacheService.upsertWatchedTxns(txStatus);
+        }
+      }
+    }
```

Comment:
This should be outside of the `if` statement as it doesn't care if there are any `extrinsicIndices` in the lot.

I'd actually suggest moving it out into a separate function as well.

```suggestion
    }
    await expirePendingTransactions(currentBlockNumber);
  }
  
  async expirePendingTransactions(currentBlockNumber: number): Promise<void> {
    // Now check all pending transactions for expiration as of this block
    const pendingTxns = await this.cacheService.getAllPendingTxns();
    
    const updateTxns: ITxStatus[] = [];
    // eslint-disable-next-line no-restricted-syntax
    for (const txStatus of Object.values(pendingTxns)) {
      if (txStatus.death <= currentBlockNumber) {
        txStatus.status = 'expired';
        updateTxns.push(txStatus);
        this.logger.verbose(`Tx ${txStatus.txHash} expired (birth: ${txStatus.birth}, death: ${txStatus.death}, currentBlock: ${currentBlockNumber})`);
      }
    }
    if (updateTxns) {
      await this.cacheService.upsertWatchedTxns(txStatus);
    }
```

Reply:
Fixed; will refactor also--upcoming